### PR TITLE
Integrate new Storage APIs for Azure Functions

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
Integrate new storage APIs for Functions.
- Add new interface `IDicomFunctionsBuilder` that extension methods can target for Azure Functions
- Refactor start up
- Remove old experimental code

## Related issues
[AB#83001](https://microsofthealth.visualstudio.com/Health/_workitems/edit/83001)

## Testing
Run unit tests. Separately, dependencies must be refactored for the indexing function

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Functions/Indexing/QueryTagIndexingOptions.cs [hunk 1]
```diff
@@ -6,14 +6,15 @@
...
+namespace Microsoft.Health.Dicom.Functions.Indexing
 {
     /// <summary>
     /// Represents the configuration for a "reindex" function.
     /// </summary>
-    public class ReindexOperationConfiguration
+    public class QueryTagIndexingOptions
     {
-        public const string SectionName = "Reindex";
+        internal const string ConfigurationSectionName = "Indexing";
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Functions/Indexing/QueryTagIndexingOptions.cs:16]: Any reason for renaming? Will it include setting for both reindexing and deindexing?

File: src/Microsoft.Health.Dicom.Functions/Management/PurgeHistoryOptions.cs [hunk 2]
```diff
@@ -9,9 +9,9 @@
 
 namespace Microsoft.Health.Dicom.Functions.Management
 {
-    public class OrchestrationHistoryConfiguration
+    public class PurgeHistoryOptions
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Functions/Management/PurgeHistoryOptions.cs:12]: use `PurgeHistoryConfiguration` for consistency?

File: src/Microsoft.Health.Dicom.Functions/Management/PurgeHistoryOptions.cs [hunk 3]
```diff
@@ -9,9 +9,9 @@
...
     {
-        public const string SectionName = "OrchestrationHistory";
+        internal const string ConfigurationSectionName = "OrchestrationHistory";
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Functions/Management/PurgeHistoryOptions.cs:14]: for consistency, `OrchestrationHistory` => `PurgeHistory` ?

File: src/Microsoft.Health.Dicom.Functions/Configuration/DicomFunctionsConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+using Microsoft.Health.Dicom.Functions.Management;
+
+namespace Microsoft.Health.Dicom.Functions.Configuration
+{
+    internal static class DicomFunctionsConfiguration
+    {
+        public const string SectionName = "DicomFunctions";
+
+        public const string PurgeFrequencyVariable = "%"
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Functions/Configuration/DicomFunctionsConfiguration.cs:14]: nit: I'm thinking should we move this constant into the job class.
So far the `xxConfiguration` includes 2 things: `SectionName` and configurations. This one is pretty special.

File: src/Microsoft.Health.Dicom.Functions/Registration/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace Microsoft.Health.Dicom.Functions.Registration
+{
+    internal static class ServiceCollectionExtensions
+    {
+        public static IServiceCollection AddFunctionsOptions<T>(this IServiceCollection services, string sectionName)
+            where T : class
+        {
+            EnsureArg.IsNotNull(services, nameof(services));
+
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Functions/Registration/ServiceCollectionExtensions.cs:22]: `EnsureArg.IsNotEmptyOrWhiteSpace(sectionName, nameof(sectionName));`?

File: src/Microsoft.Health.Dicom.Functions/Registration/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                    .GetSection(sectionName)
+                    .Bind(sectionObj));
+
+            return services;
+        }
+
+        public static IServiceCollection AddStorageServices(this IServiceCollection services, IConfiguration configuration)
+        {
+            EnsureArg.IsNotNull(services, nameof(services));
+
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Functions/Registration/ServiceCollectionExtensions.cs:37]: ensure `configuration` is not null?

**Review by pengchen0692 — APPROVED** [review]
with several minor comments.

File: src/Microsoft.Health.Dicom.Functions/Indexing/QueryTagIndexingOptions.cs (hunk 1 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions/Indexing/QueryTagIndexingOptions.cs:16]: I am trying to differentiate between things that are "configurations" and things that are "options." I can probably leave this name alone, but throughout this change I'm trying to draw this line between the data model (options) and how they're recorded (config).

File: docker/docker-compose.vs.yml [hunk 7]
```diff
@@ -18,6 +18,8 @@ services:
   functions:
     image: ${DOCKER_REGISTRY-}microsofthealthdicomfunctions
     environment:
+      # [SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine", Justification="Password is only used for development.")]
+      AzureFunctionsJobHost__SqlServer__ConnectionString: "Server=tcp:sql,1433;Initial Catalog=DICOM;Persist Security Info=False;User ID=sa;Password=${SAPASSWORD:-L0ca1P@ssw0rd};MultipleActiveResultSets=False;Connection Timeout=30;"
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply docker/docker-compose.vs.yml:22]: We have to use `AzureFunctionsJobHost` as that's built-in unfortunately. All of our config sections are prefixed by this section

File: src/Microsoft.Health.Dicom.Functions/Configuration/DicomFunctionsConfiguration.cs (hunk 4 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions/Configuration/DicomFunctionsConfiguration.cs:14]: Moved

File: src/Microsoft.Health.Dicom.Functions/Management/PurgeHistoryOptions.cs (hunk 2 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions/Management/PurgeHistoryOptions.cs:12]: In the future, I'd like to use the word "Options" to refer to these sorts of classes to divorce them conceptually from the configuration file

File: src/Microsoft.Health.Dicom.Functions/AzureFunctionsJobHost.cs [hunk 8]
```diff
@@ -7,6 +7,6 @@ namespace Microsoft.Health.Dicom.Functions
 {
     internal static class AzureFunctionsJobHost
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Functions/AzureFunctionsJobHost.cs:8]: Deleted

**Pull request merged**